File: DictionaryProject/Commands/ApplyCommand.vb
<Highlight(RegEx:="^\s*apply \\\{", Color:="Green", NextLevel:=4)> _
<CloneHighlighting(1, 4)> _
<Highlight(Regex:="\\\}", Color:="Green", Level:=4, NextLevel:=-2)> _
Public Class ApplyCommand
    Inherits CommandBase

    Private _commands As New CommandCollection
    Public ReadOnly Property Commands() As CommandCollection
        Get
            Return _commands
        End Get
    End Property

    Public Overrides Sub CheckSanity(ByVal project As Project)
        For Each cmd As CommandBase In Commands
            cmd.CheckSanity(project)
        Next
    End Sub

    Public Overrides Sub Execute(ByVal context As Context)
        Dim branchList As New List(Of Context)()
        branchList.Add(context)

        Dim branchStack As New Stack(Of IEnumerator(Of Context))
        branchStack.Push(Nothing)

        Dim branches As IEnumerator(Of Context) = context.Branches.Values.GetEnumerator()

        While branchStack.Count > 0
            While branches.MoveNext()
                branchList.Add(branches.Current)

                branchStack.Push(branches)
                branches = branches.Current.Branches.Values.GetEnumerator()
            End While

            branches = branchStack.Pop()
        End While

        For Each c As Context In branchList
            Commands.Execute(c)
        Next
    End Sub

    Public Overrides Sub LoadCommand(ByVal project As Project, ByVal reader As System.IO.TextReader, ByVal line As String, ByRef lineNumber As Integer)
        MyBase.LoadCommand(project, reader, line, lineNumber)

        Dim parts As List(Of String) = ProjectSerializer.ReadTokens(line)

        If parts.Last <> "{" Then
            project.Warnings.Add(String.Format("Line {0}: The Apply command requires a {{ at the end of the line.", lineNumber))
        Else
            parts.RemoveAt(0)
            parts.RemoveAt(parts.Count - 1)

            If parts.Count = 0 Then
                If ProjectSerializer.ReadLines(New Object() {project, Commands}, reader, AddressOf ProjectSerializer.ParseCommands, lineNumber) Then
                    project.Warnings.Add(String.Format("Line {0}: The Apply command is not closed correctly.", lineNumber))
                End If
            Else
                project.Warnings.Add(String.Format("Line {0}: The Apply command must have no arguments.", lineNumber))
            End If
        End If
    End Sub

    Public Overrides Sub WriteCommand(ByVal writer As System.IO.TextWriter)
        writer.WriteLine("Apply {")
        Commands.WriteCommands(writer)
        writer.WriteLine("}")
    End Sub
End Class

File: monowordbuilder/wordbuilderbase/Commands/ApplyCommand.cs
using System.Collections.Generic;

namespace Whee.WordBuilder.Model.Commands
{
	[Highlight(RegEx = "^\\s*apply \\\\\\{", Color = "Green", NextLevel = 4)]
	[CloneHighlighting(1, 4)]
	[Highlight(RegEx = "\\\\\\}", Color = "Green", Level = 4, NextLevel = -2)]
	public class ApplyCommand : CommandBase
	{
	
		private CommandCollection _commands = new CommandCollection();
		public CommandCollection Commands {
			get { return _commands; }
		}
		
		public override void Execute(Context context)
		{
			List<Context> branchList = new List<Context>();
			branchList.Add(context);
	
			Stack<IEnumerator<Context>> branchStack = new Stack<IEnumerator<Context>>();
			branchStack.Push(null);
	
			IEnumerator<Context> branches = context.Branches.Values.GetEnumerator();
	
			while (branchStack.Count > 0) {
				while (branches.MoveNext()) {
					branchList.Add(branches.Current);
	
					branchStack.Push(branches);
					branches = branches.Current.Branches.Values.GetEnumerator();
				}
	
				branches = branchStack.Pop();
			}
	
			foreach (Context c in branchList) {
				Commands.Execute(c);
			}
		}

        public override void LoadCommand(Whee.WordBuilder.ProjectV2.IProjectSerializer serializer)
        {
            m_lineNumber = serializer.LineNumber;

            ProjectV2.CommandBlockNode cbn = new ProjectV2.CommandBlockNode(serializer);

            if (cbn.Commands.Count > 0)
            {
                Children.Add(cbn);

                foreach (CommandBase cmd in cbn.Commands)
                {
                    Commands.Add(cmd);
                }
            }
            else
            {
                serializer.Warn("The apply command expects a command block.", this);
            }
        }

		public override void LoadCommand(Project project, System.IO.TextReader reader, string line, ref int lineNumber)
		{
			base.LoadCommand(project, reader, line, ref lineNumber);
	
			List<string> parts = ProjectSerializer.ReadTokens(line);
	
			if (parts[parts.Count - 1] != "{") {
				project.Warnings.Add(string.Format("Line {0}: The Apply command requires a {{ at the end of the line.", lineNumber));
			}
			else {
				parts.RemoveAt(0);
				parts.RemoveAt(parts.Count - 1);
	
				if (parts.Count == 0) {
					if (ProjectSerializer.ReadLines(new object[] { project, Commands }, reader, ProjectSerializer.ParseCommands, ref lineNumber)) {
						project.Warnings.Add(string.Format("Line {0}: The Apply command is not closed correctly.", lineNumber));
					}
				}
				else {
					project.Warnings.Add(string.Format("Line {0}: The Apply command must have no arguments.", lineNumber));
				}
			}
		}
	
		public override void WriteCommand(System.IO.TextWriter writer)
		{
			writer.WriteLine("Apply {");
			Commands.WriteCommands(writer);
			writer.WriteLine("}");
		}

        public override void CheckSanity(Project project, Whee.WordBuilder.ProjectV2.IProjectSerializer serializer)
        {
            foreach (CommandBase cmd in Commands)
            {
                cmd.CheckSanity(project, serializer);
            }
        }

        public override bool RequireNewLine
        {
            get
            {
                return false;
            }
        }
    }
}